apps/prairielearn/src/ee/auth/lti13/lti13Auth.ts

```diff
@@ -0,0 +1,302 @@
+      // Reasonable default is "name"
+      // Points back to OIDC Standard Claims https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims
+      name = _get(lti13_claims, lti13_instance.name_attribute);
+    }
+
+    const userInfo = {
+      uid,
+      uin,
+      name,
+      provider: 'LTI 1.3',
+      institution_id: lti13_instance.institution_id,
+    };
+
+    // AUTHENTICATE
+    await authnLib.loadUser(req, res, userInfo);
+
+    // Record the LTI 1.3 user's subject id
+    await queryAsync(sql.update_lti13_users, {
+      user_id: res.locals.authn_user.user_id,
+      lti13_instance_id: lti13_instance.id,
+      sub: lti13_claims.sub,
+    });
+
+    // Get the target_link out of the LTI request and redirect
+    const redirUrl =
+      lti13_claims['https://purl.imsglobal.org/spec/lti/claim/target_link_uri'] || '/pl';
+    res.redirect(redirUrl);
+  }),
+);
+
+export default router;
+
+//
+// Schema to validate OIDC, LTI
+//
+const OIDCAuthResponseSchema = z.object({
+  state: z.string(),
+  id_token: z.string(),
+  // also has utf8, authenticity_token, lti_storage_target
+});
+
+const OIDCLaunchFlowSchema = z.object({
+  iss: z.string(),
+  login_hint: z.string(),
+  lti_message_hint: z.string().optional(),
+  lti_deployment_id: z.string().optional(),
+  client_id: z.string().optional(),
+  target_link_uri: z.string(),
+  // also has deployment_id, canvas_environment, canvas_region, lti_storage_target
+});
+
+// Validate LTI 1.3
+// https://www.imsglobal.org/spec/lti/v1p3#required-message-claims
+const LTI13Schema = z.object({
+  'https://purl.imsglobal.org/spec/lti/claim/message_type': z.literal('LtiResourceLinkRequest'),
+  'https://purl.imsglobal.org/spec/lti/claim/version': z.literal('1.3.0'),
+  'https://purl.imsglobal.org/spec/lti/claim/deployment_id': z.string(),
+  'https://purl.imsglobal.org/spec/lti/claim/target_link_uri': z.string(),
+  'https://purl.imsglobal.org/spec/lti/claim/resource_link': z.object({
+    id: z.string(),
+    description: z.string().nullish(),
+    title: z.string().nullish(),
+  }),
+  // https://www.imsglobal.org/spec/security/v1p0/#tool-jwt
+  // https://www.imsglobal.org/spec/security/v1p0/#id-token
+  iss: z.string(),
+  aud: z.string(),
+  sub: z.string(),
+  exp: z.number(),
+  iat: z.number(),
+  azp: z.string().optional(),
+  nonce: z.string(),
+
+  given_name: z.string().optional(),
+  family_name: z.string().optional(),
+  name: z.string().optional(),
+  email: z.string().optional(),
+  locale: z.string().optional(),
+  // Could be more from OIDC Standard Claims
+  'https://purl.imsglobal.org/spec/lti/claim/roles': z.string().array(),
+
+  'https://purl.imsglobal.org/spec/lti/claim/context': z
+    .object({
+      id: z.string(),
+      type: z.string().array().optional(),
+      label: z.string().optional(),
+      title: z.string().optional(),
+    })
+    .nullable()
+    .optional(),
```

Comment:
AFAICT the spec doesn't define whether "optional" consistently means "present but maybe null" or "maybe not present at all". I've observed both of these in your sample Canvas response, so I'm erring on the side of caution and using `.nullish()`

```suggestion
    .nullish(),
```